File: GUI.cs
using System;
using System.Text;
using Juh.List;

namespace Practice
{
    public class GUI
    {
        private static readonly int[] COL_WIDTHS = {2, 30, 5, 5};
        private Charset[] _charsets;
        private int _curSet;
        private DoublyLinkedList<Student> List;

        private Charset Chars => _charsets[_curSet % _charsets.Length];

        public void Start(DoublyLinkedList<Student> list = null)
        {
            _charsets = new[]
            {
                new Charset('─', '│',
                    '┌', '┬', '┐',
                    '├', '┼', '┤',
                    '└', '┴', '┘'),
                new Charset('-', '|', '+'),
                new Charset(' ', '─', ' ',
                    ' ', '│', ' ',
                    ' ', '│', ' ',
                    '─', '┼', '─',
                    ' ', '│', ' '),
                new Charset(' ', '-', ' ',
                    ' ', '|', ' ',
                    ' ', '|', ' ',
                    '-', '+', '-',
                    ' ', '|', ' ')
            };

            if (list != null) List = list;
            else List = new DoublyLinkedList<Student>();

            Console.Clear();
            WriteListContent(list);
            WaitForInput();
        }

        public void WaitForInput()
        {
            Console.WriteLine("Input a character");
            Console.WriteLine("1: Add element.");
            Console.WriteLine("2: Remove element.");
            Console.WriteLine("3: Get element.");
            Console.WriteLine("4: Show Filtered List.");
            Console.WriteLine("5: Show Reverse list.");
            Console.WriteLine("6: Close the program.");
            Console.WriteLine("0: Change border style.");

            var ch = '\u0000';
            var num = -1;
            do
            {
                var key = Console.ReadKey();
                ch = key.KeyChar;
            } while (!int.TryParse(ch.ToString(), out num) || num > 6);

            Console.Clear();
            WriteListContent(List);

            switch (num)
            {
                case 1:
                    Add();
                    break;
                case 2:
                    Remove();
                    break;
                case 3:
                    Get();
                    break;
                case 4:
                    Filtered();
                    break;
                case 5:
                    Reverse();
                    break;
                case 6:
                    break;
                case 0:
                    _curSet++;
                    Console.Clear();
                    WriteListContent(List);
                    WaitForInput();
                    break;
            }
        }

        private void Add()
        {
            Console.WriteLine("Adding new student");
            var name = ReadString("Input name");
            var grade = ReadNumber("Input grade");
            var wasInArmy = ReadBool("Was in army?");
            Console.WriteLine("");

            var student = new Student(name, grade, wasInArmy);
            var listCopy = new DoublyLinkedList<Student>(List);
            listCopy.Add(student);

            ListPrompt(listCopy, replaceText: "Add new element?",
                visualList: new DoublyLinkedList<Student>(student));
        }

        private void Remove()
        {
            Console.WriteLine("Removing student");
            var id = ReadNumber("Input ID of element to remove");
            var listCopy = new DoublyLinkedList<Student>(List);
            try
            {
                listCopy.Delete(id);

                ListPrompt(listCopy, replaceText: "Delete this element?",
                    visualList: new DoublyLinkedList<Student>(List.Get(id)));
            }
            catch (IndexOutOfRangeException)
            {
                Cancel("Index out of range. ");
            }
        }

        private void Get()
        {
            Console.WriteLine("Getting element");
            var id = ReadNumber("Input ID of element to show");
            Student student;
            try
            {
                student = List.Get(id);

                WriteListContent(new DoublyLinkedList<Student>(student));
                ListPrompt(text: "");
            }
            catch (IndexOutOfRangeException)
            {
                Cancel("Index out of range. ");
            }
        }

        private void Filtered()
        {
            Console.WriteLine("Filtering list");
            ListPrompt(new DoublyLinkedList<Student>(List.Filter()));
        }

        private void Reverse()
        {
            Console.WriteLine("Reversing");
            var amount = ReadNumber("Input amount of elements to include into reversed array");
            try
            {
                ListPrompt(List.FromLasts(amount));
            }
            catch (ArgumentOutOfRangeException)
            {
                if (amount <= 0) Cancel("Too small amount of elements specified. ");
                else Cancel("too large amount is specified. ");
            }
        }

        private void Cancel(string reason)
        {
            ListPrompt(text: reason + "Operation canceled. ");
        }

        private void ListPrompt(DoublyLinkedList<Student> list = null,
            string text = "Changes Applied. ", string replaceText = "Replace current list with a new one?",
            DoublyLinkedList<Student> visualList = null)
        {
            visualList ??= list;

            if (Console.CursorLeft != 0) Console.WriteLine("");
            if (list! != null)
            {
                WriteListContent(visualList);
                if (ReadBool(replaceText))
                {
                    List = list;
                    Console.WriteLine("");
                }
                else
                {
                    Cancel("");
                }
            }

            Console.WriteLine(text + "Press any key to continue.");
            Console.ReadKey();
            Console.Clear();
            WriteListContent(List);
            WaitForInput();
        }

        private bool ReadBool(string prompt)
        {
            ClearLine();
            Console.WriteLine(prompt);
            Console.Write("Press ");
            Console.ForegroundColor = ConsoleColor.Green;
            Console.Write("y");
            Console.ResetColor();
            Console.Write(" for yes or ");
            Console.ForegroundColor = ConsoleColor.Red;
            Console.Write("n");
            Console.ResetColor();
            Console.Write(" for no: ");

            char ch;
            var pos = Console.CursorLeft;
            do
            {
                if (Console.CursorLeft != pos)
                {
                    Console.CursorLeft = pos;
                    Console.Write(" ");
                    Console.CursorLeft = pos;
                }

                var key = Console.ReadKey();
                ch = key.KeyChar;

                switch (ch)
                {
                    case 'y':
                    case 'Y':
                    case '\r':
                        return true;
                    case 'n':
                    case 'N':
                        return false;
                }
            } while (true);
        }

        private int ReadNumber(string prompt)
        {
            ClearLine();
            int num;
            Console.Write(prompt + ": ");
            var text = Console.ReadLine();
            while (!int.TryParse(text, out num))
            {
                Console.CursorTop--;
                ClearLine();
                Console.Write("Invalid number. " + prompt + ": ");
                text = Console.ReadLine();
            }

            return num;
        }

        private string ReadString(string prompt)
        {
            ClearLine();
            Console.Write(prompt + ": ");
            var text = Console.ReadLine();
            while (text == null || (text = text.Trim()).Length == 0)
            {
                Console.CursorTop--;
                ClearLine();
                Console.Write("Input string empty. " + prompt + ": ");
                text = Console.ReadLine();
            }

            return text;
        }

        private void ClearLine()
        {
            var curLine = Console.CursorTop;
            Console.CursorLeft = 0;
            Console.Write(new string(' ', Console.WindowWidth));
            Console.SetCursorPosition(0, curLine);
        }

        private string FormatString(string name, int length, bool right = false)
        {
            if (name.Length > length)
            {
                name = name.Substring(0, length - 4);
                name += "...";
            }
            else
            {
                name = right ? name.PadLeft(length) : name.PadRight(length);
            }

            return name;
        }

        private string SplitterLine(char[] line)
        {
            return line[0] + new string(line[3], COL_WIDTHS[0] + 2) +
                   line[1] + new string(line[3], COL_WIDTHS[1] + 2) +
                   line[1] + new string(line[3], COL_WIDTHS[2] + 2) +
                   line[1] + new string(line[3], COL_WIDTHS[3] + 2) + line[2];
        }

        private string TextLine(string idCol, string nameCol, string gradeCol, string armyCol)
        {
            return Chars.LeftVertical + $" {FormatString(idCol, COL_WIDTHS[0], true)} " +
                   Chars.CentralVertical + $" {FormatString(nameCol, COL_WIDTHS[1])} " +
                   Chars.CentralVertical + $" {FormatString(gradeCol, COL_WIDTHS[2], true)} " +
                   Chars.CentralVertical + $" {FormatString(armyCol, COL_WIDTHS[3])} " + Chars.RightVertical;
        }

        private void WriteListContent(DoublyLinkedList<Student> list)
        {
            var text = new StringBuilder();
            text.AppendLine(SplitterLine(Chars.TopRow()));
            text.AppendLine(TextLine("Id", "Name", "Grade", "Army"));
            text.AppendLine(SplitterLine(Chars.MiddleRow()));

            var id = 0;

            foreach (var student in list)
                text.AppendLine(TextLine((id++).ToString(), student.Name, student.Grade.ToString(),
                    student.Army.ToString()));

            text.AppendLine(SplitterLine(Chars.BottomRow()));

            Console.WriteLine(text.ToString());
        }
    }

    public struct Charset
    {
        public Charset(char topHorizontal, char centralHorizontal, char bottomHorizontal,
            char leftVertical, char centralVertical, char rightVertical,
            char topLeftCorner, char topIntersection, char topRightCorner,
            char leftIntersection, char centralIntersection, char rightIntersection,
            char bottomLeftCorner, char bottomIntersection, char bottomRightCorner)
        {
            TopHorizontal = topHorizontal;
            CentralHorizontal = centralHorizontal;
            BottomHorizontal = bottomHorizontal;
            LeftVertical = leftVertical;
            CentralVertical = centralVertical;
            RightVertical = rightVertical;
            TopLeftCorner = topLeftCorner;
            TopIntersection = topIntersection;
            TopRightCorner = topRightCorner;
            LeftIntersection = leftIntersection;
            CentralIntersection = centralIntersection;
            RightIntersection = rightIntersection;
            BottomLeftCorner = bottomLeftCorner;
            BottomIntersection = bottomIntersection;
            BottomRightCorner = bottomRightCorner;
        }

        public Charset(char horizontal, char vertical, char topLeftCorner, char topIntersection, char topRightCorner,
            char leftIntersection, char centralIntersection, char rightIntersection, char bottomLeftCorner,
            char bottomIntersection, char bottomRightCorner) :
            this(horizontal, horizontal, horizontal,
                vertical, vertical, vertical,
                topLeftCorner, topIntersection, topRightCorner,
                leftIntersection, centralIntersection, rightIntersection,
                bottomLeftCorner, bottomIntersection, bottomRightCorner)
        {
        }

        public Charset(char horizontal, char vertical, char intersection) : this(horizontal, vertical,
            intersection, intersection, intersection,
            intersection, intersection, intersection,
            intersection, intersection, intersection)
        {
        }

        public Charset(char topHorizontal, char centralHorizontal, char bottomHorizontal,
            char leftVertical, char centralVertical, char rightVertical, char intersection) :
            this(topHorizontal, centralHorizontal, bottomHorizontal,
                leftVertical, centralVertical, rightVertical,
                intersection, intersection, intersection,
                intersection, intersection, intersection,
                intersection, intersection, intersection)
        {
        }

        public char[] TopRow()
        {
            return TableArray()[0];
        }

        public char[] MiddleRow()
        {
            return TableArray()[1];
        }

        public char[] BottomRow()
        {
            return TableArray()[2];
        }

        public char[][] TableArray()
        {
            return new[]
            {
                new[] {TopLeftCorner, TopIntersection, TopRightCorner, TopHorizontal},
                new[] {LeftIntersection, CentralIntersection, RightIntersection, CentralHorizontal},
                new[] {BottomLeftCorner, BottomIntersection, BottomRightCorner, BottomHorizontal}
            };
        }

        public char TopHorizontal { get; }
        public char CentralHorizontal { get; }
        public char BottomHorizontal { get; }
        public char LeftVertical { get; }
        public char CentralVertical { get; }
        public char RightVertical { get; }
        public char TopLeftCorner { get; }
        public char TopIntersection { get; }
        public char TopRightCorner { get; }
        public char LeftIntersection { get; }
        public char CentralIntersection { get; }
        public char RightIntersection { get; }
        public char BottomLeftCorner { get; }
        public char BottomIntersection { get; }
        public char BottomRightCorner { get; }
    }
}

File: src/cf/juh9870/GUI.java
package cf.juh9870;

import cf.juh9870.lists.Student;
import cf.juh9870.lists.StudentsList;

public class GUI {
	private static final int[] COL_WIDTHS = {2, 30, 14, 12};
	private Charset[] charsets;
	private int curSet = 0;
	private StudentsList list;

	public static String padRight(String s, int n) {
		return String.format("%-" + n + "s", s);
	}

	public static String padLeft(String s, int n) {
		return String.format("%" + n + "s", s);
	}

	public Charset getChars() {
		return charsets[curSet % charsets.length];
	}

	public void start(StudentsList list) {
		charsets = new Charset[]
				{
						new Charset('─', '│',
								'┌', '┬', '┐',
								'├', '┼', '┤',
								'└', '┴', '┘'),
						new Charset('-', '|', '+'),
						new Charset(' ', '─', ' ',
								' ', '│', ' ',
								' ', '│', ' ',
								'─', '┼', '─',
								' ', '│', ' '),
						new Charset(' ', '-', ' ',
								' ', '|', ' ',
								' ', '|', ' ',
								'-', '+', '-',
								' ', '|', ' ')
				};

		if (list != null) this.list = list;
		else this.list = new StudentsList();

		Console.start();

		Console.clear();
		writeListContent(list);
		waitForInput();
	}

	public void waitForInput() {
		Console.printLine("Input a character");
		Console.printLine("1: Add element.");
		Console.printLine("2: Remove element.");
		Console.printLine("3: Get element.");
		Console.printLine("4: Show Filtered List.");
		Console.printLine("5: Show Reverse list.");
		Console.printLine("6: Close the program.");
		Console.printLine("0: Change border style.");

		char ch = '\u0000';
		int num = -1;
		do {
			ch = Console.getKeyPress();
			try {
				num = Integer.parseInt(Character.toString(ch));
			} catch (NumberFormatException ignored) {

			}
		} while (num < 0 || num > 6);

		Console.clear();
		writeListContent(list);

		switch (num) {
			case 1:
				add();
				break;
			case 2:
				remove();
				break;
			case 3:
				get();
				break;
			case 4:
				filtered();
				break;
			case 5:
				reverse();
				break;
			case 6:
				break;
			case 0:
				curSet++;
				Console.clear();
				writeListContent(list);
				waitForInput();
				break;
		}
	}

	private void add() {
		Console.printLine("Adding new student");
		String name = readString("Input name");
		int monthBorn = readNumber("Input month of birth");
		int dayBirth = readNumber("Input day of birth");
		Console.printLine("");

		Student student = new Student(name, dayBirth, monthBorn);
		StudentsList listCopy = new StudentsList(list);
		listCopy.add(student);

		listPrompt(listCopy, null, "Add new element?", new StudentsList(student));
	}

	private void remove() {
		Console.printLine("Removing student");
		int id = readNumber("Input ID of element to remove");
		StudentsList listCopy = new StudentsList(list);
		try {
			listCopy.delete(id);

			listPrompt(listCopy, null, "Delete this element?", new StudentsList(list.get(id)));
		} catch (IndexOutOfBoundsException e) {
			cancel("Index out of range. ");
		}
	}

	private void get() {
		Console.printLine("Getting element");
		int id = readNumber("Input ID of element to show");
		Student student;
		try {
			student = list.get(id);

			writeListContent(new StudentsList(student));
			listPrompt(null, "", null, null);
		} catch (IndexOutOfBoundsException e) {
			cancel("Index out of range. ");
		}
	}

	private void filtered() {
		Console.printLine("Filtering list");
		listPrompt(new StudentsList(list.filter()), null, null, null);
	}

	private void reverse() {
		Console.printLine("Reversing");
		int amount = readNumber("Input amount of elements to include into reversed array");
		try {
			listPrompt(list.fromLasts(amount), null, null, null);
		} catch (IllegalArgumentException e) {
			if (amount <= 0) cancel("Too small amount of elements specified. ");
			else cancel("too large amount is specified. ");
		}
	}

	private void cancel(String reason) {
		listPrompt(null, reason + "Operation canceled. ", null, null);
	}

	private void listPrompt(StudentsList list,
							String text, String replaceText,
							StudentsList visualList) {
		if (text == null) text = "Changes Applied. ";
		if (replaceText == null) replaceText = "Replace current list with a new one?";
		if (visualList == null) visualList = list;

		Console.printLine("");
		if (list != null) {
			writeListContent(visualList);
			if (readBool(replaceText)) {
				this.list = list;
				Console.printLine("");
			} else {
				cancel("");
			}
		}

		Console.printLine(text + "Press Enter to continue.");
		Console.getKeyPress();
		Console.clear();
		writeListContent(this.list);
		waitForInput();
	}

	private boolean readBool(String prompt) {
		Console.printLine(prompt);
		Console.print("Press " + Console.Color.GREEN + "y" + Console.Color.RESET +
				" or " + Console.Color.GREEN + "enter" + Console.Color.RESET + " for yes, or "
				+ Console.Color.RED + "n" + Console.Color.RESET + " for no: ");

		char ch;
		do {
			ch = Console.getKeyPress();

			switch (ch) {
				case 'y':
				case 'Y':
				case '\r':
					return true;
				case 'n':
				case 'N':
					return false;
			}
		} while (true);
	}

	private int readNumber(String prompt) {
		int num = -1;
		Console.printLine(prompt + ": ");
		String text = Console.readLine();
		try {
			num = Integer.parseUnsignedInt(text);
		} catch (NumberFormatException e) {
			while (num < 0) {
				Console.printLine("Invalid number. " + prompt + ": ");
				text = Console.readLine();
				try {
					num = Integer.parseUnsignedInt(text);
				} catch (NumberFormatException ignored) {
				}
			}
		}

		return num;
	}

	private String readString(String prompt) {
		Console.printLine(prompt + ": ");
		String text = Console.readLine();
		while (text == null || (text = text.trim()).length() == 0) {
			Console.printLine("Input String empty. " + prompt + ": ");
			text = Console.readLine();
		}

		return text;
	}

	private String repeat(char c, int amout) {
		return new String(new char[amout]).replace('\0', c);
	}

	private String formatString(String name, int length, boolean right) {
		if (name.length() > length) {
			name = name.substring(0, length - 4);
			name += "...";
		} else {
			name = right ? padLeft(name, length) : padRight(name, length);
		}

		return name;
	}


	private String splitterLine(char[] line) {
		return line[0] + repeat(line[3], COL_WIDTHS[0] + 2) +
				line[1] + repeat(line[3], COL_WIDTHS[1] + 2) +
				line[1] + repeat(line[3], COL_WIDTHS[2] + 2) +
				line[1] + repeat(line[3], COL_WIDTHS[3] + 2) + line[2] + "\n";
	}

	private String textLine(String idCol, String nameCol, String monthCol, String dayCol) {
		return getChars().LeftVertical + " " + formatString(idCol, COL_WIDTHS[0], true) + " " +
				getChars().CentralVertical + " " + formatString(nameCol, COL_WIDTHS[1], false) + " " +
				getChars().CentralVertical + " " + formatString(monthCol, COL_WIDTHS[2], true) + " " +
				getChars().CentralVertical + " " + formatString(dayCol, COL_WIDTHS[3], false) + " " +
				getChars().RightVertical + "\n";
	}

	private void writeListContent(StudentsList list) {
		StringBuilder text = new StringBuilder();
		text.append(splitterLine(getChars().TopRow()));
		text.append(textLine("Id", "Name", "Month of Birth", "Day of Birth"));
		text.append(splitterLine(getChars().MiddleRow()));

		int id = 0;

		for (Student student : list)
			text.append(textLine(Integer.toString(id++), student.name,
					student.monthName(),
					Integer.toString(student.birthDay)));

		text.append(splitterLine(getChars().BottomRow()));

		Console.printLine(text.toString());
	}

	public static class Charset {
		public char TopHorizontal;
		public char CentralHorizontal;
		public char BottomHorizontal;
		public char LeftVertical;
		public char CentralVertical;
		public char RightVertical;
		public char TopLeftCorner;
		public char TopIntersection;
		public char TopRightCorner;
		public char LeftIntersection;
		public char CentralIntersection;
		public char RightIntersection;
		public char BottomLeftCorner;
		public char BottomIntersection;
		public char BottomRightCorner;

		public Charset(char topHorizontal, char centralHorizontal, char bottomHorizontal,
					   char leftVertical, char centralVertical, char rightVertical,
					   char topLeftCorner, char topIntersection, char topRightCorner,
					   char leftIntersection, char centralIntersection, char rightIntersection,
					   char bottomLeftCorner, char bottomIntersection, char bottomRightCorner) {
			TopHorizontal = topHorizontal;
			CentralHorizontal = centralHorizontal;
			BottomHorizontal = bottomHorizontal;
			LeftVertical = leftVertical;
			CentralVertical = centralVertical;
			RightVertical = rightVertical;
			TopLeftCorner = topLeftCorner;
			TopIntersection = topIntersection;
			TopRightCorner = topRightCorner;
			LeftIntersection = leftIntersection;
			CentralIntersection = centralIntersection;
			RightIntersection = rightIntersection;
			BottomLeftCorner = bottomLeftCorner;
			BottomIntersection = bottomIntersection;
			BottomRightCorner = bottomRightCorner;
		}

		public Charset(char horizontal, char vertical, char topLeftCorner, char topIntersection, char topRightCorner,
					   char leftIntersection, char centralIntersection, char rightIntersection, char bottomLeftCorner,
					   char bottomIntersection, char bottomRightCorner) {
			this(horizontal, horizontal, horizontal,
					vertical, vertical, vertical,
					topLeftCorner, topIntersection, topRightCorner,
					leftIntersection, centralIntersection, rightIntersection,
					bottomLeftCorner, bottomIntersection, bottomRightCorner);

		}

		public Charset(char horizontal, char vertical, char intersection) {
			this(horizontal, vertical,
					intersection, intersection, intersection,
					intersection, intersection, intersection,
					intersection, intersection, intersection);

		}

		public Charset(char topHorizontal, char centralHorizontal, char bottomHorizontal,
					   char leftVertical, char centralVertical, char rightVertical, char intersection) {
			this(topHorizontal, centralHorizontal, bottomHorizontal,
					leftVertical, centralVertical, rightVertical,
					intersection, intersection, intersection,
					intersection, intersection, intersection,
					intersection, intersection, intersection);

		}

		public char[] TopRow() {
			return TableArray()[0];
		}

		public char[] MiddleRow() {
			return TableArray()[1];
		}

		public char[] BottomRow() {
			return TableArray()[2];
		}

		public char[][] TableArray() {
			return new char[][]
					{
							new char[]{
									TopLeftCorner, TopIntersection, TopRightCorner, TopHorizontal
							},
							new char[]{
									LeftIntersection, CentralIntersection, RightIntersection, CentralHorizontal
							},
							new char[]{
									BottomLeftCorner, BottomIntersection, BottomRightCorner, BottomHorizontal
							}
					};
		}
	}
}

